packages/flutter/lib/src/widgets/two_dimensional_viewport.dart

```diff
@@ -1304,23 +1356,59 @@ abstract class RenderTwoDimensionalViewport extends RenderBox implements RenderA
 
   void _insertChild(RenderBox child, ChildVicinity slot) {
     assert(_debugTrackOrphans(newOrphan: _children[slot]));
+    assert(!_keepAliveBucket.containsValue(child));
     _children[slot] = child;
     adoptChild(child);
   }
 
   void _moveChild(RenderBox child, {required ChildVicinity from, required ChildVicinity to}) {
-    if (_children[from] == child) {
-      _children.remove(from);
+    final TwoDimensionalViewportParentData childParentData = parentDataOf(child);
+    if (!childParentData.keptAlive) {
+      if (_children[from] == child) {
+        _children.remove(from);
+      }
+      assert(_debugTrackOrphans(newOrphan: _children[to], noLongerOrphan: child));
+      _children[to] = child;
+      return;
+    }
+    // If the child in the bucket is not current child, that means someone has
+    // already moved and replaced current child, and we cannot remove this
+    // child.
+    if (_keepAliveBucket[childParentData.vicinity] == child) {
+      _keepAliveBucket.remove(childParentData.vicinity);
     }
-    assert(_debugTrackOrphans(newOrphan: _children[to], noLongerOrphan: child));
-    _children[to] = child;
+    assert(() {
+      _debugDanglingKeepAlives.remove(child);
+      return true;
+    }());
+    // If there is an existing child in the new slot, that mean that child
+    // will be moved to other index. In other cases, the existing child should
+    // have been removed by _removeChild. Thus, it is ok to overwrite it.
+    assert(() {
+      if (_keepAliveBucket.containsKey(childParentData.vicinity)) {
+        _debugDanglingKeepAlives.add(_keepAliveBucket[childParentData.vicinity]!);
+      }
+      return true;
+    }());
+    _keepAliveBucket[childParentData.vicinity] = child;
   }
 
   void _removeChild(RenderBox child, ChildVicinity slot) {
-    if (_children[slot] == child) {
-      _children.remove(slot);
+    final TwoDimensionalViewportParentData childParentData = parentDataOf(child);
+    if (!childParentData.keptAlive) {
+      if (_children[slot] == child) {
+        _children.remove(slot);
+      }
+      assert(_debugTrackOrphans(noLongerOrphan: child));
+      dropChild(child);
+      return;
     }
-    assert(_debugTrackOrphans(noLongerOrphan: child));
+    assert(_keepAliveBucket[childParentData.vicinity] == child);
```

Comment:
if they are in _debugDanglingKeepAlives they won't be in here, no?

Reply:
They should be. Every child in _debugDanglingKeepAlives should be in the _keepAliveBucket, but not necessarily the other way around.

Reply:
you are right, I misread the code